### ISSUE TITLE
Code Cell Output Improvements (labels, captions, ...)

### ISSUE THREAD

**Issue by chrisjsewell** [executablebooks/MyST-NB#72]
We would like to, at a minimum, allow for code outputs such as plots and tables to
have a label and caption. Then also e.g. the other options allowed by [docutils figures](https://docutils.sourceforge.io/docs/ref/rst/directives.html#figure).

Using pasting and gluing (see b46c2f1f5b8d155866e531a1199188fd588517bd and #70) actually gives us a lot more flexibility in this respect, to add these options with markdown cell text.
It is basically an implementation of the Model-View-Controller pattern that I have banged on about before, so I would rather use this approach.

But for 'simpler' use cases we may also want to allow for these to be set in the metadata of the code cell, under a suitable key name:

```json
{
  "output_format": {
    "label": "fig:myfig",
    "caption": "blah blah blah",
    "width": 400
}
```

**Comment by chrisjsewell** [comment]
Note a caveat/shortcoming of the cell metadata approach, is that you have to assume that the cell only has one output, which this metadata would be applied to. This is what is done in IPyPublish: https://ipypublish.readthedocs.io/en/latest/metadata_tags.html#cell-output-level

**Comment by choldgraf** [comment]
Imo we could start with a few of the most common artifacts one would put in a document (tables and figures seem like a good start) and build roles or directives around each.

**Comment by choldgraf** [comment]
This also makes me think about tangle js:

http://worrydream.com/Tangle/

I wonder if that can be inspiration for the future

**Comment by akhmerov** [comment]
IMO this approach has a danger of bundling too much into the cell metadata.

An alternative approach would be to implement figures as containers, which would include complete cells inside them.

**Comment by chrisjsewell** [comment]
I think the original intention of this issue was closed in #243 and https://myst-nb.readthedocs.io/en/latest/use/formatting_outputs.html#images
Although, I can certainly think of more improvements to add in the future 😄 
Fell free to open another issue, or re0open with a more specific title/goal
